src/lock.rs

```diff
@@ -0,0 +1,96 @@
+use std::fs::{self, read_to_string, write, File};
+use std::io::{BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use md5::Context;
+use serde::{Deserialize, Serialize};
+
+use crate::util::Fallible;
+
+pub fn code_changed<P: AsRef<Path>>(lock_file: P, code_dir: P) -> std::io::Result<bool> {
+    if code_dir.as_ref().exists()
+        && code_dir.as_ref().is_dir()
+        && lock_file.as_ref().exists()
+        && lock_file.as_ref().is_file()
+    {
+        let mut code_hasher = Hasher::default();
+        code_hasher.consume_all(code_dir)?;
+        let code_hash = code_hasher.compute();
+
+        let lock = Lock::read(lock_file)?;
+
+        Ok(lock.code_hash != code_hash)
+    } else {
+        Ok(false)
+    }
+}
+
+/// Content for Yarner.lock files
+#[derive(Serialize, Deserialize)]
+pub struct Lock {
+    pub source_hash: String,
+    pub code_hash: String,
+}
+
+impl Lock {
+    pub fn read<P: AsRef<Path>>(path: P) -> std::io::Result<Self> {
+        let buf = read_to_string(path)?;
+        let val = toml::from_str::<Self>(&buf)?;
+
+        Ok(val)
+    }
+
+    pub fn write<P: AsRef<Path>>(&self, path: P) -> Fallible {
+        let str = toml::to_string(self)?;
+        write(path, str)?;
+
+        Ok(())
+    }
+}
+
+pub struct Hasher {
+    hasher: Context,
+}
+
+impl Default for Hasher {
+    fn default() -> Self {
+        Self {
+            hasher: Context::new(),
+        }
+    }
+}
+
+impl Hasher {
+    pub fn consume_all<P: AsRef<Path>>(&mut self, path: P) -> std::io::Result<()> {
```

Comment:
It seems like you hash only file contents but not paths. So if I do not change the content but rename a file, the hash does not change?

Related to this, I suspect the contents of `Yarner.toml` should probably be included for both source and code as the options can influence the output as well?

Reply:
Thanks for pointing that out! I will definitely include paths.

For the `Yarner.toml` I am not really sure. Changes here could break...

* Reverse mode, if code labels change
* The document (e.g. name or macro syntax changes)

Also, if we hash that file in both directions, does that not lock everything once I change it? So you need to `--force` after any change...

Reply:
> Also, if we hash that file in both directions, does that not lock everything once I change it?

I was coming from incremental computation, i.e. only recompute something if the inputs have changed. But you want to use this in the opposite direction, do not do something if the output has been changed since the last computation. So not including `Yarner.toml` should be fine because it isn't overwritten in one direction or the other.

Reply:
(Hashing paths might actually be weird because on `path::Path` cannot be accessed as `&[u8]` on non-Unix platforms. Probably `.as_os_str().to_string_lossy()` or `.to_str().ok_or("I really do not process non-UTF-8 paths my friend.")`...)

Reply:
This point seems moot if individual paths and hashes are listed...